File: music-api/frontend/src/types/release.ts
import type { Music } from './music'

export type ReleaseType =
  | 'SINGLE'
  | 'EP'
  | 'ALBUM'

export type Release = {
  id: number
  artist_id: number
  title: string
  release_type: ReleaseType
  cover_image_url: string
  cover_image_public_id: string
  description: string
  release_date: string | null
  is_published: boolean
  created_at: string
  updated_at: string
}

export type ReleaseDetails = {
  release: Release
  tracks: Music[]
}

export type CreateReleaseInput = {
  title: string
  release_type: ReleaseType
  cover_image_url: string
  cover_image_public_id: string
  description: string
  release_date: string
  is_published: boolean
}

export type UpdateReleaseInput =
  CreateReleaseInput

export type AddReleaseTrackInput = {
  track_number: number
}

File: music-api/internal/models/release.go
package models

import "time"

// ReleaseType identifies the kind of music release.
type ReleaseType string

const (
	ReleaseTypeSingle ReleaseType = "SINGLE"
	ReleaseTypeEP     ReleaseType = "EP"
	ReleaseTypeAlbum  ReleaseType = "ALBUM"
)

// Release represents a collection of music published by an artist.
//
// A SINGLE, EP, or ALBUM is represented by the same domain model.
// Individual tracks remain stored in the music table.
type Release struct {
	ID int `json:"id"`

	ArtistID int `json:"artist_id"`

	Title string `json:"title"`

	ReleaseType ReleaseType `json:"release_type"`

	CoverImageURL string `json:"cover_image_url"`

	CoverImagePublicID string `json:"cover_image_public_id"`

	Description string `json:"description"`

	ReleaseDate *time.Time `json:"release_date"`

	IsPublished bool `json:"is_published"`

	CreatedAt time.Time `json:"created_at"`

	UpdatedAt time.Time `json:"updated_at"`
}

// ReleaseDetails is the public representation of a release together
// with its ordered tracks.
type ReleaseDetails struct {
	Release Release `json:"release"`

	Tracks []Music `json:"tracks"`
}
